internal/app/synchronizer/synchronizer_service.go

```diff
@@ -46,6 +132,165 @@ func (s *synchronizerService) Register(ctx context.Context, req *pb.RegisterRequ
+	// requests that the results are ready.
+	<-s.cycleData.resultsReady
+	results = make([]*pb.Match, len(s.cycleData.requests[id].results))
+	copy(results, s.cycleData.requests[id].results)
+	synchronizerServiceLogger.Debugf("Results ready for id: %v, results: %v, elapsed time: %v", id, getMatchIds(results), time.Since(t).String())
+	return results
+}
+
+// Evaluate aggregates all the proposals across requests and calls the user configured
+// evaluator with these. Once evaluator returns, it copies results back in respective
+// request data objects and then signals the completion of evaluation. It then waits
+// for all the result processing (by fetchMatches) to be compelted before changing the
+// synchronizer state to idle.
+// NOTE: This method is always called while holding the synchronizer state mutex.
+func (s *synchronizerService) Evaluate() {
+	var aggregateProposals []*pb.Match
+	proposalMap := make(map[string]string)
+	for id, data := range s.cycleData.requests {
+		aggregateProposals = append(aggregateProposals, data.proposals...)
+		for _, m := range data.proposals {
+			proposalMap[m.MatchId] = id
+		}
+	}
+
+	synchronizerServiceLogger.Infof("Requesting evaluation of %v proposals", len(aggregateProposals))
+	// TODO: Errors in evaluation are currently ignored. This should be handled and should lead to
+	// error being surfaced to all the pending requests to fetch matches.
+	results := s.evalFunc(aggregateProposals)
+	synchronizerServiceLogger.Infof("Evaluation returned %v results", len(results))
+	for _, m := range results {
+		cid, ok := proposalMap[m.MatchId]
+		if !ok {
+			// Evaluator returned a match that the synchronizer did not submit for evaluation. This
+			// could happen if the match id was modified accidentally.
+			synchronizerServiceLogger.Warningf("Result %v does not belong to any known requests.", m)
+			continue
+		}
+
+		s.cycleData.requests[cid].results = append(s.cycleData.requests[cid].results, m)
+	}
+
+	// Signal completion of evaluation to all requests waiting for results.
+	close(s.cycleData.resultsReady)
+
+	// Wait for all fetchMatches to complete processing results and then set synchronizer
+	// to idle state and signal any blocked registration requests to proceed.
+	s.cycleData.pendingRequests.Wait()
+	synchronizerServiceLogger.Infof("Changing state from evaluating to idle")
+	s.state = idle
+	s.idleCond.Broadcast()
+}
+
+func (s *synchronizerService) trackRegistrationWindow() {
+	time.Sleep(s.registrationInterval())
+	s.stateMutex.Lock()
+	defer s.stateMutex.Unlock()
+	synchronizerServiceLogger.Infof("Changing state from requestRegistration to proposalCollection")
+	s.state = proposalCollection
+	go s.trackProposalWindow()
+}
+
+func (s *synchronizerService) trackProposalWindow() {
+	time.Sleep(s.proposalCollectionInterval())
+	s.stateMutex.Lock()
+	defer s.stateMutex.Unlock()
+	synchronizerServiceLogger.Infof("Changing state from proposalCollection to evaluation")
+	s.state = evaluation
+	s.Evaluate()
```

Comment:
One of my concerns is since we don't know what user would do in the evaluator function, we may have a customized evaluator function that takes more than `s.proposalCollectionInterval` to complete, which results in more and more `Resgiter` calls waiting for the lock and ends up with pod crashes or the evaluator takes too long to respond. Do you think it is worth to consider? Would it be possible to create a new cycleData object per cycle, and execute `s.evalFunc` in a gorountine? This way we could have multiple `s.Evaluate()` running in parallel without hurting the atomicity

Reply:
You bring up a good point. See similar comment from Jeremy below. Custom Evaluator taking long would block all pending requests - thereby queueing up any new registrations. I would be concerned with having eval functions run in paralell though - since correctness wise, it would not be right to start registrations when an eval is pending. The timeout suggestion jeremy had might solve this problem - and we can abandon the pending evaluation request. Have created an issue to address this. Thanks!